### ISSUE TITLE
Cannot impl Drop for a generic type with a Higher-Rank Trait Bound

### ISSUE THREAD

**Issue by dwrensha** [rust-lang/rust#27863]
If I try to compile this program:

``` rust
pub trait Foo<'a> {}

pub struct Bar<T> where T: for<'a> Foo<'a> {
    _phantom: ::std::marker::PhantomData<T>,
}

impl <T> Drop for Bar<T> where T: for<'a> Foo<'a> {
    fn drop(&mut self) {}
}

fn main() {}
```

I get this clearly bogus error message:

```
$ rustc test.rs
test.rs:7:1: 9:2 error: The requirement `for<'a> T : Foo<'a>` is added only by the Drop impl. [E0367]
test.rs:7 impl <T> Drop for Bar<T> where T: for<'a> Foo<'a> {
test.rs:8     fn drop(&mut self) {}
test.rs:9 }
test.rs:7:1: 9:2 help: run `rustc --explain E0367` to see a detailed explanation
test.rs:3:1: 5:2 note: The same requirement must be part of the struct/enum definition
test.rs:3 pub struct Bar<T> where T: for<'a> Foo<'a> {
test.rs:4     _phantom: ::std::marker::PhantomData<T>,
test.rs:5 }
```

I tried fiddling with it to get it to typecheck, but nothing worked.

```
$ rustc --version --verbose
rustc 1.4.0-nightly (e822a18ae 2015-08-16)
binary: rustc
commit-hash: e822a18ae7d55cefc332c6598a607cef0554ec77
commit-date: 2015-08-16
host: x86_64-apple-darwin
release: 1.4.0-nightly
```

**Comment by alexcrichton** [comment]
cc @pnkfelix

**Comment by Boiethios** [comment]
Same here, I just stumbled on this issue. How is this doing? The issue is there since 3 years.

**Comment by Ralith** [comment]
Affecting me also. Bump?

**Comment by cjgillot** [comment]
Original report now correctly compiles.
